Add cppstd as a subsetting of compiler

jgsogo opened the pull request:
Changelog: Feature: Add `compiler.cppstd` setting (mark `cppstd` as deprecated)
Docs: https://github.com/conan-io/docs/pull/1266

This PR includes changes from PR #4986 and #4942 (really short ones)

----

Some hints about the new/old behavior:
 - It keeps previous behavior: all use cases (tests) should be summarized here https://github.com/conan-io/conan/pull/4986, 
 - The default value (or `None`) for `compiler.cppstd` generates the same ID as it was generating for `cppstd`.
 - If the user specifies `cppstd` and `compiler.cppstd`, Conan will raise. Yes, even if those have the same values.
 - Package ID for non-default values generates different ID for `cppstd` and `compiler.cppstd`
 - If no value is provided to `compiler.cppstd` then the default value will be assigned to it.

Summarizing:
```
package_id("cppstd=default") == package_id("cppstd=None")
package_id("cppstd=default") == package_id("compiler.cppstd=default")
package_id("compiler.cppstd=default") == package_id("compiler.cppstd=None")
package_id("cppstd=20") != package_id("compiler.cppstd=20")
```

----

closes #4873 

@TAGS: slow

jgsogo commented on conans/model/info.py:
```diff
@@ -397,15 +397,16 @@ def default_std_matching(self):
-        if self.full_settings.cppstd and \
-                self.full_settings.compiler and \
-                self.full_settings.compiler.version:
+        if self.full_settings.compiler and \
```
Do we want to maintain this `default` behavior?

jgsogo commented on conans/test/functional/configuration/compiler_cppstd_test.py:
```diff
@@ -0,0 +1,101 @@
+        self.assertIn("Package 'hh': The specified 'compiler.cppstd=14' and 'cppstd=11'"
+                      " are different", self.t.out)
+
+    @unittest.expectedFailure
```
In this test, although the recipe does not define `settings = ... "compiler"`, the settings object contains that value, it detects the mismatch and raises.

jgsogo commented:
Green!

![image](https://media.giphy.com/media/YFkpsHWCsNUUo/giphy.gif)

jgsogo commented on conans/model/conan_file.py:
```diff
@@ -145,6 +152,13 @@ def initialize(self, settings, env):
         except ConanException:
             pass
 
+        try:
```
Need to try/except?

memsharded commented on conans/model/conan_file.py:
```diff
@@ -132,6 +133,12 @@ def initialize(self, settings, env):
         self.requires = create_requirements(self)
         self.settings = create_settings(self, settings)
+
+        try:
```
I think this check might be redundant:
- It is not possible to get cppstd defined if it doesn't come from the profile
- The profile was checked before being applied to conanfiles

If the problem is per-package settings, I think I'd try to implement that in the same place as the global cppstd is checked, to keep that logic together.

jgsogo replied:
Yes, this is about checking the _per-package settings_. 

I tried to do it (there are some ideas in [this PR](https://github.com/jgsogo/conan/pull/11) which I moved from this repo to mine), but there is a problem: **I need to constrain the settings for the package before checking them** and that action is performed inside this function, in the call to `self.settings = create_settings(self, settings)`.

I would really like to move this check to previous stages (this is executed once the graph starts to run), but it would require to move too many things not related to this PR, don't you think so?

lasote replied:
I don't get it. I will go ask you "live" LoL

memsharded commented on conans/test/functional/cppstd/default_cppstd_test.py:
```diff
@@ -0,0 +1,182 @@
+    def run(self, *args, **kwargs):
+        default_profile_path = os.path.join(temp_folder(), "default.profile")
+        save(default_profile_path, self.default_profile)
+        with environment_append({"CONAN_DEFAULT_PROFILE_PATH": default_profile_path}):
```
This looks complicated, why not directly writing the home/profiles/default profile?

jgsogo replied:
Why? It is a documented way to override the default profile: _"[TIP:] using the environment variable `CONAN_DEFAULT_PROFILE_PATH`"_. I think this is the easiest way to be sure that every `TestClient` created in these tests will use the same profile...

uilianries commented on conans/client/settings_preprocessor.py:
```diff
@@ -1,30 +1,73 @@
+                             " Use only the former one.")
+
+    if cppstd:
+        warnings.warn("Setting 'cppstd' is deprecated in favor of 'compiler.cppstd'")
```
I'm curious, why not `self.output.warn`?

jgsogo replied:
This raises a `UserWarning` if you have configured _warnings as errors_ (we do in the test suite), if not it just prints the message.

lasote commented on conans/client/migrations.py:
```diff
@@ -154,6 +154,8 @@ def _make_migrations(self, old_version):
         if old_version < Version("1.15.0"):
             migrate_registry_file(self.cache, self.out)
 
+        # TODO: Add settings migration
```
I don't think we should. The current policy is just to fast forward if not changed. Related with #5038

jgsogo replied:
Understood, didn't know about that issue/PR.

lasote commented on conans/model/info.py:
```diff
@@ -398,14 +401,36 @@ def default_std_matching(self):
+        self._std_matching = None  # Do nothing
+
+    # Handle std_matching strategy for compiler.cppstd
+    _std_matching = None
```
This variable here feels weird, why is it needed here and not at `__init__`?

lasote replied:
Maybe it reads easier if the var is always a function (empty lambda by default?) and it is called differently, something like: `self._adjust_settings_for_std`. You wouldn't need to write `if self._std_matching` just call always the `self._adjust_settings_for_std`.

lasote commented on conans/test/functional/configuration/profile_test.py:
```diff
@@ -551,6 +551,7 @@ def profile_crazy_inheritance_test(self):
                              [settings]
                              arch=x86_64
                              compiler=Visual Studio
+                             compiler.cppstd=14
```
Why this change?

jgsogo replied:
Now we are setting `compiler.cppstd` to the default value if the user is not providing one (check function `_fill_compiler_cppstd`) although we remove it for ID computation.

lasote commented on conans/test/functional/cppstd/default_cppstd_test.py:
```diff
@@ -34,14 +33,15 @@ class Library(ConanFile):
                 cppstd = self.settings.get_safe("cppstd")
-                self.output.info(">>>> settings: {{}}".format(self.settings.fields))
-                self.output.info(">>>> cppstd: {{}}".format(cppstd))
+                compiler_cppstd = self.settings.get_safe("compiler.cppstd")
```
Do we want to alter this test? Or keep it to ensure we don't break anything and do another one for the new subsetting?

jgsogo replied:
I've minimized changes in this file, now I think it is clear that the modifications (additions) don't affect existing behavior. Tell me what you think, thanks!

memsharded commented on conans/model/info.py:
```diff
@@ -398,14 +404,31 @@ def default_std_matching(self):
+        self._adjust_settings_for_std = lambda u: u  # Do nothing
+
+    def _settings_sha(self):
+        settings = self.settings.copy()
```
I suggest:
- Having a default adjust = None
- If adjust = None, no need to copy, no need to call it.
- The adjust would do the copy internally, and return the sha. In this way it is not possible to forget to copy, which would destroy the ``info`` original values.

jgsogo replied:
I agree with the `adjust=None`, but I prefer to leave the `copy` and `sha` outside the _adjust_ function, maybe we can add here more _adjustments_ only for computing the ID and this way we can create the pipeline (I'm not sure if it will make sense right now for the `build_settings` and `toolset` ones):

```python
    def _settings_sha(self):
        if self._adjust_settings_for_std:
            settings = self.settings.copy()
            settings = self._adjust_settings_for_std(settings)
            settings = self._adjust_other_thing(settings)
            settings = self._adjust_...(settings)
            settings = ....
            return settings.sha
        return self.settings.sha
```

Just tell me if you really think it is better to make the copy inside the function and return the sha 😉

memsharded commented on conans/model/conan_file.py:
```diff
@@ -145,6 +147,10 @@ def initialize(self, settings, env):
             pass
 
+        if 'cppstd' in self.settings.fields:
+            self.output.writeln("Setting 'cppstd' is deprecated in favor of 'compiler.cppstd',"
```
Why not a warning? It will be more visible, and help spotting the message

jgsogo commented:
Relevant changes since last reviews:
 * **Use `compiler.cppstd` in build_helpers**, until now I wasn't taking into account `compiler.cppstd` in the build helpers 😓. I've written a helper function `cppstd_from_settings`, but not sure if [`conans/client/build/cppstd_flags.py`](https://github.com/conan-io/conan/pull/4917/files#diff-0ae047f956f76dcfd983a59666652cfb) is the place to define it.
 * _Pure C_ recipes will need to remove `compiler.cppstd` (as they were already removing `compiler.libcxx`) or Conan will inject the compiler flag for the C++ standard and some compilers will raise an error
    * So the autogenerated `conanfile.py` for testing has a different checksum.

memsharded commented:
What happens to existing pure C libraries that are already removing ``compiler.libcxx``, but will not be removing the default added ``compiler.cppstd``? Will the ID be affected?

jgsogo commented:
> What happens to existing pure C libraries that are already removing `compiler.libcxx`, but will not be removing the default added `compiler.cppstd`? Will the ID be affected?

The problem is not the ID (it won't be affected because the `compiler.cppstd=<default>` is not used to compute the ID), *the problem is that those libraries could fail* if the compiler raises for the _new_ flag... 

![image](https://user-images.githubusercontent.com/1406456/56977828-679a7700-6b76-11e9-96aa-661ec2749dcc.png)

...and I'm not sure if we can handle this use case, as we can't know that those libraries are _pure C_. Probably this is a **breaking change** ⚠️ ⚠️ ⚠️

jgsogo commented:
Thinking about this _breaking change_ a little bit deeper:
 * This flag is being introduced in the following elements:
   * `AutoToolsBuildEnvironment`: this _new_ flag goes to `CXXFLAGS`
   * `Meson`: goes to `... -D cpp_std="c++14"`. And that option is different from `c_std` ([docs](http://mesonbuild.com/Builtin-options.html)). 
   * `CMake`: populates:
      ```
      set(CONAN_CMAKE_CXX_STANDARD 14)
      set(CONAN_CMAKE_CXX_EXTENSIONS ON/OFF)
      set(CONAN_STD_CXX_FLAG c++14)
      ...
      set(CMAKE_CXX_FLAGS "${CONAN_STD_CXX_FLAG} ${CMAKE_CXX_FLAGS}")  # CMake <= 3.1
      set(CMAKE_CXX_STANDARD ${CONAN_CMAKE_CXX_STANDARD})
      set(CMAKE_CXX_EXTENSIONS ${CONAN_CMAKE_CXX_EXTENSIONS})
      ```
   * `VisualStudioBuildEnvironment`: goes to dict entry `"CL": "... /std:c++14 ..."` which is then used by `MSBuild` and `VirtualBuildEnvGenerator`.
   * `CompilerArgsGenerator`: it goes directly to the list of flags

So, I think that the problem could be with `CompilerArgsGenerator` (it was the test that failed and I commented before), but all the other ones should avoid this flag when compiling a C library.

lasote commented on conans/client/migrations_settings.py:
```diff
@@ -288,3 +288,75 @@
 settings_1_14_3 = settings_1_14_2
 settings_1_14_4 = settings_1_14_3
+
+settings_1_15_0 = """
```
I think this might be better to be adjusted when the release is merged to develop probably. In that point, we have "freeze" the settings for the version, not before. Why did you add it?

jgsogo commented:
I need to modify docs according to the _new_ behavior
